Incoming avoidance

liamaxelrod opened the pull request:
This will finish off the milestone of incoming avoidance since it is now fulfilled all criteria

So I'm making a pull request to 
close #22 

### :ballot_box_with_check: Definition of Done checklist
- [x] Meaningful title and valuable description in pull request
- [x] Code change tested manually if not covered by automated tests
- [x] Pull request linked to one or more issues/stories
- [x] Acceptance criteria of the linked stories satisfied
- [x] Code change comprehended by those who approved it
- [x] State of target branch maintained or improved after the code change

liamaxelrod committed:
added the logic for the safety feature button

Pretty much I just made a movie on that when the safety button is
pressed it will turn false. Causing all the safety features to
turn off for the simulated car.

liamaxelrod committed:
implemented backup and forward incoming avoidance

So the simulator car will now backup or move forward
if the infrared picks up an optical at least 20 distance from it.
Also I added a bouillon that I quoted out in case it needs to be used.
this commitment fulfills the criteria and issue #22 that are
1,2,3,4

liamaxelrod committed:
Merge branch 'master' into incoming_avoidance

liamaxelrod committed:
Merge branch 'master' into incoming_avoidance

Valkama1 committed:
update variable name

liamaxelrod committed:
implemented different levels in incoming avoidance

pretty much now when the car gets too close to an object
there are different levels of thresholds. The closer the object
the faster the car will retreat or move forward to get away from
that object.

This is all attached to issue #22
fulfilling criteria 5

AlbinKarlsson commented on arduino/smartcar/smartcar.ino:
```diff
@@ -11,8 +11,11 @@
 MQTTClient mqtt;
 WiFiClient net;
 
+bool safetyFeatures = false;
```
Add a comment, what this variable does/represent

liamaxelrod replied:
Implemented comment

AlbinKarlsson commented on arduino/smartcar/smartcar.ino:
```diff
@@ -133,10 +136,16 @@ void loop()
             const auto backIRDis = backIR.getDistance();
             
-            if (safetySystem){ //check if the safety system is enabled
+
```
You removed a useful comment here, maybe replace it with another comment.

liamaxelrod replied:
Added the comment

AlbinKarlsson committed:
Comments

There was an important comment above the lastWill method that was deleted. I put it back at its rightful place

AlbinKarlsson commented on arduino/smartcar/smartcar.ino:
```diff
@@ -155,7 +164,68 @@ void loop()
+
+void incomingAvoidanceThreshold(long frontUltDis, long frontIRDis, long backIRDis)
+{
+    if (frontUltDis <= 30 && frontUltDis != 0 || frontIRDis <= 40 && frontIRDis != 0)//Ford threshold 1
```
Ford threshold 1? Maybe add more details to this comment so it is easier to understand. 
Add more to the other if else conditionals too.

AlbinKarlsson commented on arduino/smartcar/smartcar.ino:
```diff
@@ -155,7 +164,68 @@ void loop()
     
 }
 
-//If the connection breaks, this method will be called
```
You had removed one of my comments and put your code between my comment and my method. Just be careful when you add code that it doesn't affect other methods/functions.

liamaxelrod replied:
I see that but at the end I did put a new comment with your last well what method

AlbinKarlsson commented on arduino/smartcar/smartcar.ino:
```diff
@@ -155,7 +164,68 @@ void loop()
+    } 
+}
+
+//This method will be called when the connection breaks from the broker 
```
I added my comment back here above my method

liamaxelrod replied:
Thank you I appreciate that

AlbinKarlsson commented on arduino/smartcar/smartcar.ino:
```diff
@@ -155,7 +164,68 @@ void loop()
+                car.setSpeed(0);
+                Serial.println("Emergency stop 1");
+            }
+            canDrive = false;//so the car can move in the stop soon
```
You wrote soon here, I assume you mean Zone ;)

liamaxelrod replied:
Thank you for noticing that

AlbinKarlsson commented on arduino/smartcar/smartcar.ino:
```diff
@@ -155,7 +164,68 @@ void loop()
+        car.setSpeed(-60);
+        Serial.println("backing up level 2");
+        activeAvoidance = true;
+    } else if (frontUltDis <= 90 && frontUltDis != 0 || frontIRDis <= 40 && frontIRDis != 0)//Ford threshold 3this
```
At the end of you comment here you wrote 3this. Not sure what you wanted to write here. 
Add more details like i mentioned in the previous comment

liamaxelrod replied:
It's fixed now

AlbinKarlsson reviewed (changes requested):
I have added some comments. The code looks good, however I noticed a few things. Please revise! :D

liamaxelrod committed:
Update smartcar.ino

I went over the comments Albin made when reviewing my pull request and fix and corrected them as needed

liamaxelrod commented:
There's been a couple of problems during this pull request so I'm gonna start fresh from a new pull request
